ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Codec.hs

```diff
@@ -77,10 +79,7 @@ codecBlockFetchUnwrapped encodeBlock     decodeBlock
                                           <$> decodeBlock
       (ServerAgency TokStreaming, 5) -> return $ SomeMessage MsgBatchDone
 
-      -- TODO proper exceptions
-      (ClientAgency TokIdle, _)      -> fail "codecBlockFetch.Idle: unexpected key"
-      (ServerAgency TokBusy, _)      -> fail "codecBlockFetch.Busy: unexpected key"
-      (ServerAgency TokStreaming, _) -> fail "codecBlockFetch.Streaming: unexpected key"
+      _ -> fail (printf "codecBlockFetch (%s) unexpected key (%d, %d)" (show stok) key len)
```

Comment:
For these cases, can we do both? Lets keep the exhaustive pattern matching, rather than introducing a wildcard. But lets also have a function in a `where` clause for the error message, and call that for each case.

Reply:
Right, this will protect us if we add a state in the future.